overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
 	}
 	return seededSys, nil
 }
+
+func createSystemForModelFromValidatedSnaps(getInfo func(name string) (*snap.Info, bool, error), db asserts.RODatabase, label string, model *asserts.Model) (newFiles []string, dir string, err error) {
+	if isUC20 := model.Grade() != asserts.ModelGradeUnset; !isUC20 {
+		return nil, "", fmt.Errorf("cannot create a system for non UC20 model")
+	}
+
+	logger.Noticef("creating recovery system with label %q for %q", label, model.Model())
+
+	// TODO: should that path provided by boot package instead?
+	recoverySystemDirInRootDir := filepath.Join("/systems", label)
+	assertedSnapsDir := filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps")
+	recoverySystemDir := filepath.Join(boot.InitramfsUbuntuSeedDir, recoverySystemDirInRootDir)
+
+	wOpts := &seedwriter.Options{
+		// RW mount of ubuntu-seed
+		SeedDir: boot.InitramfsUbuntuSeedDir,
+		Label:   label,
+	}
+	w, err := seedwriter.New(model, wOpts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	optsSnaps := make([]*seedwriter.OptionsSnap, 0, len(model.RequiredWithEssentialSnaps()))
+	// collect all snaps that are present
+	modelSnaps := make(map[string]*snap.Info)
+
+	getModelSnap := func(name string, essential bool, nonEssentialPresence string) error {
+		kind := "essential"
+		if !essential {
+			kind = "non-essential"
+			if nonEssentialPresence != "" {
+				kind = fmt.Sprintf("non-essential but %q",
+					nonEssentialPresence)
+			}
+		}
+		info, present, err := getInfo(name)
+		if err != nil {
+			return fmt.Errorf("cannot obtain %v snap information: %v", kind, err)
+		}
+		if !essential && !present && nonEssentialPresence == "optional" {
+			// non-essential snap which is declared as optionally
+			// present in the model
+			return nil
+		}
+		// grab those
+		logger.Noticef("%v snap: %v", kind, name)
+		if !present {
+			return fmt.Errorf("internal error: %v snap %q not present", kind, name)
+		}
+		if _, ok := modelSnaps[info.MountFile()]; ok {
+			// we've already seen this snap
+			return nil
+		}
+		// present locally
+		optsSnaps = append(optsSnaps, &seedwriter.OptionsSnap{
+			Path: info.MountFile(),
+		})
```

Comment:
in the case of grade: dangerous here we could specify a channel which is not the model channel, it's a bit unclear whether we should simply take the current channel or something else?  a TODO would be good about this though

Reply:
added a TODO about this